# ML Zoomcamp 2023 - Homework #1

Name: Wong Chee Fah

Email: wongcheefah@gmail.com

In [1]:
import pandas as pd
import numpy as np

### Question 1: Version of Pandas

In [2]:
pd.__version__

'2.0.3'

#### Ans: 2.0.3

### Question 2: Number of columns in the dataset

In [3]:
housing = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv')
print(housing.shape)
housing.info()

(20640, 10)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


#### Ans: 10

### Question 3: Select columns with missing data

In [4]:
housing.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

#### Ans: total_bedrooms

### Question 4: Number of unique values in the 'ocean_proximity' column

In [5]:
housing['ocean_proximity'].nunique()

5

#### Ans: 5

### Question 5: Average value of the 'median_house_value' for the houses near the bay

In [6]:
housing[housing['ocean_proximity']=='NEAR BAY']['median_house_value'].mean()

259212.31179039303

#### Ans: 259212

### Question 6: Has the mean value changed after filling missing values?

In [7]:
ttl_bdrm_mean = housing['total_bedrooms'].mean()
ttl_bdrm_mean

537.8705525375618

In [8]:
housing['total_bedrooms'].fillna(ttl_bdrm_mean).mean()

537.8705525375617

#### Ans: No

### Questions 7: Value of the last element of w

In [9]:
X = housing[housing['ocean_proximity']=='ISLAND'][['housing_median_age', 'total_rooms', 'total_bedrooms']].values
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [10]:
XTX = X.T.dot(X)
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [11]:
inv_XTX = np.linalg.inv(XTX)
inv_XTX

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [12]:
y = np.array([950, 1300, 800, 1000, 1300])
y

array([ 950, 1300,  800, 1000, 1300])

In [13]:
w = inv_XTX.dot(X.T).dot(y)
w

array([23.12330961, -1.48124183,  5.69922946])

#### Ans: 5.6992